### ISSUE TITLE
Support a broader range of versions of the underlying OCaml compiler

### ISSUE THREAD

**Issue by jchavarri** [melange-re/melange#886]
Currently the approach Melange is using for every newly released version is designed to work with one version of OCaml, and just one. The OCaml compiler version chosen is the latest one available (5.1 right now).

The reason for this is that Melange has its own [fork of OCaml compiler libs](https://github.com/melange-re/melange-compiler-libs), and it's not trivial to support multiple versions of the OCaml compiler without breaking tooling like [odoc](https://github.com/melange-re/melange/issues/603) or merlin/lsp. So supporting a single version guarantees that the tooling that relies on artifacts generated by Melange like `cmt` files doesn't break when trying to read them with a different version of the OCaml compiler.

This constraint is too tight though. We are starting to see problems across the ecosystem that manifest in different ways:
- Some OCaml libraries have a hard time accepting PRs to support Melange, because it requires either constraining the `ocaml` version too much, or releasing two separate packages, see an example in https://github.com/anuragsoni/routes/pull/151.
- Some projects supporting Melange have examples that force them to run their CI only on the latest version of the OCaml compiler, see an example in https://github.com/melange-community/melange-webapi/issues/14.

The way I see it, there are two paths for this:
- Update odoc and merlin/lsp so that they read cmts using the Melange compiler libs functions, not the "stock" OCaml compiler ones
- Try to modify Melange so that it includes multiple versions of the compiler libs, and dynamically switches to the one installed, kind of like a "ppxlib" for compiler libs.

Any other approaches that I'm missing?

**Comment by EduardoRFS** [comment]
For smaller versions CMT can be migrated down, https://github.com/EduardoRFS/ocaml-migrate-types is outdated but the concept works.

**Comment by anmonteiro** [comment]
if https://github.com/ocaml/ocaml/pull/12734 merges, we might be able to get back a wider range of OCaml versions for compilation, but not for editor integration.

**Comment by jchavarri** [comment]
Thanks for the feedback.

> For smaller versions CMT can be migrated down, https://github.com/EduardoRFS/ocaml-migrate-types is outdated but the concept works.

This is very interesting! I understand with a solution like this, Melange would be able to read which major version of the OCaml compiler is being used in the current switch and produce CMT files that are compatible with that version, so that tooling like odoc and merlin doesn't break?

**Comment by jchavarri** [comment]
Another use case for solving this issue —besides the 2 mentioned on the description— is the following: if more versions of the underlying OCaml compiler were supported, users could install the latest versions of Melange in projects where there is a backend written in native OCaml and frontend in Melange, and for whatever reason the backend has to stay in version 5.0 or earlier. This issue is affecting Ahrefs heavily, up to the point where we had to create our own fork of Melange to keep up with the additions included in the latest versions of Melange while staying compatible with 4.14.

**Comment by anmonteiro** [comment]
There are really big downsides to migrating the typedtree, though:

- now we need to include however many versions of compiler-libs we support
  - makes compilation quite slower
  - makes melange executable size a lot bigger
- extra burden of migrating the typechecker when a new version of OCaml comes out, before we even try to migrate to that version
- the maintenance on `ocaml-migrate-types` has been inconsistent at best, so it's unclear who'd own that piece

**Comment by jchavarri** [comment]
> makes compilation quite slower

Which compilation, the `melange` package compilation itself or the actual compilation of `cmt` files and other artifacts generated by `melc`? The former is a one-time cost on installation, so I understand we could afford some slow down in that case.

Also, how much slower? Is it 1%, 100%?

> makes melange executable size a lot bigger

Same question, how bigger? Right now it's:
```
$ ls -lah _opam/bin/melc
-rwxr-xr-x 1 me me 31M Nov  7 14:56 _opam/bin/melc
```

Would it be a few KB more, or a lot of MB more?

> extra burden of migrating the typechecker when a new version of OCaml comes out, before we even try to migrate to that version
> the maintenance on `ocaml-migrate-types` has been inconsistent at best, so it's unclear who'd own that piece

If we decide to adopt this, maybe I could commit to maintain `ocaml-migrate-types` when new versions of OCaml come out. I think that's a better use of my time than maintaining a branch of Melange that is compatible with 4.14, or answering to feedback from users due to the narrow support for OCaml, which is where my time is being spent at the moment. Although, it could be that maintaining `ocaml-migrate-types` takes longer than those tasks 😅 

> Right now these downsides don't justify the benefit that it'd bring for me. Do you disagree?

I haven't formed an opinion yet, but I am convinced it's not ideal that Ahrefs can't follow the latest versions of Melange. It's bad for both Ahrefs —we can't get latest fixes and improvements without investing time on updating our fork— and Melange —we're not able to test new releases internally before publishing them. One could say it's a temporary situation, and everything will be fixed when Ahrefs upgrades to 5.1, but the same will happen again when 5.2 is released: as soon as Melange upgrades to that version, we'd be stuck again.

**Comment by jchavarri** [comment]
> One could say it's a temporary situation, and everything will be fixed when Ahrefs upgrades to 5.1

I took a look at how fast Ahrefs would adopt new versions of the OCaml compiler in the past (<5.x), and it seems upgrades would happen just a few weeks (or even days) after a release. So it could be that the upgrade to OCaml 5 is really exceptional.

**Comment by anmonteiro** [comment]
This was fixed in https://github.com/melange-re/melange/pull/1028 and #1071 additionally adds support for 5.2
